state/upgrades.go

```diff
@@ -283,10 +280,25 @@ func EnsureApplicationCharmOriginsHaveRevisions(pool *StatePool) error {
 			if curl.Revision == -1 {
 				return errors.Errorf("charm url %q has no revision", curl.String())
 			}
+
+			if charm.Local.Matches(curl.Schema) {
```

Comment:
If the origin.source is good, which we've seen to be true,  we should use that, not the curl.

Reply:
Unsure what you mean here

I don't think we do know origin.source is good here. If we switch from a charmhub to a local charm using pylibjuju, this check will pass because the charm url will point to a local charm, but the origin will still be an old charm-hub one

But even if the origin.source was correct, such as in the case of switching from a charmhub to a local charm using the juju CLI, the rest of the origin will not be updated (id/hash left behind, revsion incorrect), so I don't see what we'd change?